Use ivalue::Future in autograd engine and DistEngine.

pritamdamania87 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #43684 Preserve python backtrace in autograd engine errors.
* **#43676 Use ivalue::Future in autograd engine and DistEngine.**

This is one part of https://github.com/pytorch/pytorch/issues/41574 to
ensure we consolidate everything around ivalue::Future.

I've removed the use of torch/csrc/utils/future.h from the autograd engines and
used ivalue::Future instead.

Differential Revision: [D23362415](https://our.internmc.facebook.com/intern/diff/D23362415/)

pritamdamania committed:
Use ivalue::Future in autograd engine and DistEngine.

This is one part of https://github.com/pytorch/pytorch/issues/41574 to
ensure we consolidate everything around ivalue::Future.

I've removed the use of torch/csrc/utils/future.h from the autograd engines and
used ivalue::Future instead.

Differential Revision: [D23362415](https://our.internmc.facebook.com/intern/diff/D23362415/)

[ghstack-poisoned]

albanD commented on torch/csrc/autograd/engine.cpp:
```diff
@@ -454,8 +454,8 @@ void GraphTask::mark_as_completed_and_run_post_processing() {
-    // waitNoThrow() to guarantee the future is marked complete on exit.
-    future_result_->waitNoThrow();
+    // wait() to guarantee the future is marked complete on exit.
+    future_result_->wait();
```
Is this one never throwing? Shall we handle that?

pritamdamania87 replied:
This one doesn't throw, since the ivalue::Future API is slightly different. `wait()` doesn't return the value and it is just void and `value()` actually returns the value and throws an exception.

albanD replied:
Sounds good!

albanD commented on torch/csrc/autograd/engine.cpp:
```diff
@@ -871,7 +871,12 @@ auto Engine::execute(const edge_list& roots,
 
-  return execute_with_graph_task(graph_task, graph_root)->wait();
+  execute_with_graph_task(graph_task, graph_root);
+  // Avoid a refcount bump for the Future, since we check for refcount in
```
I'm not sure what this comment refers to? Why does it talk about the distributed engine?

pritamdamania87 replied:
It is referring to this check in DistEngine: https://github.com/pytorch/pytorch/blob/master/torch/csrc/distributed/autograd/engine/dist_engine.cpp#L559. If we incur a refcount bump here, that check would fail.

albanD replied:
Maybe refer to that line or a PR talking about this because I don't think it is clear otherwise :/

pritamdamania87 replied:
I updated the PR to include some more details about this check in DistEngine referring to the line we're talking about here.

albanD reviewed (commented):
Looks quite good.
I guess that in these APIs, functions with the same name do the same thing right?

pritamdamania committed:
Update on "Use ivalue::Future in autograd engine and DistEngine."


This is one part of https://github.com/pytorch/pytorch/issues/41574 to
ensure we consolidate everything around ivalue::Future.

I've removed the use of torch/csrc/utils/future.h from the autograd engines and
used ivalue::Future instead.

Differential Revision: [D23362415](https://our.internmc.facebook.com/intern/diff/D23362415/)

[ghstack-poisoned]

albanD reviewed (approved):
Looks good!
Thanks
